lib/linked_list.rb

```diff
@@ -18,88 +18,227 @@ def initialize
+      until current.nil?
+        return true if current.data == value
+        current = current.next
+      end
+
+      return false
     end
 
     # method to return the max value in the linked list
     # returns the data value and not the node
+    # Time Complexity: O(n) - traversing through all of the nodes
+    # Space Complexity: O(1)
     def find_max
-      raise NotImplementedError
+      # what if there's no nodes/empty list? return nil?
+      return nil if @head.nil?
+
+      max = @head.data
+      current = @head
+      until current.nil?
+        max = [max, current.data].max
+        current = current.next
+      end
+
+      return max
     end
 
     # method to return the min value in the linked list
     # returns the data value and not the node
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def find_min
-      raise NotImplementedError
+      return nil if @head.nil?
+
+      min = @head.data
+      current = @head
+
+      until current.nil?
+        min = current.data if current.data < min
+        current = current.next
+      end
+
+      return min
     end
 
 
     # method that returns the length of the singly linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def length
-      raise NotImplementedError
+      count = 0
+      current = @head
+
+      until current.nil?
+        count += 1
+        # move the pointer to the next node
+        current = current.next
+      end
+
+      return count
     end
 
     # method that returns the value at a given index in the linked list
     # index count starts at 0
-    # returns nil if there are fewer nodes in the linked list than the index value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # returns nil if there are fewer nodes in the linked list than the index value or if the list is empty
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def get_at_index(index)
-      raise NotImplementedError
+      return nil if self.length < index || @head.nil?
+
+      # need to consider edge case where index == 0, return @head.data
+      current = @head
+      count = 0
+      until count == index
+        count += 1
+        current = current.next
+      end
+
+      # count is equal to index
+      return current.data
+
     end
 
     # method to print all the values in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def visit
-      raise NotImplementedError
+      # what happens if linked list is empty?
+
+      return nil if @head.nil?
+
+      current = @head
+      until current.nil?
+        puts current.data
+        current = current.next
+      end
+
     end
 
     # method to delete the first node found with specified value
     # Time Complexity: ?
     # Space Complexity: ?
     def delete(value)

```

Comment:
👍 , however this is a little over-complicated.  

1.  You don't need the `else` in line 145 because the if has a return.
2. Similar issues with the if head is the node to delete.